### PR TITLE
Sum-check protocol

### PR DESCRIPTION
**Pull request by Al-Kindi-0**
## Describe your changes
Implements the sum-check protocol for use in the virtual bus as in #1182 .

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            let reduced_evaluation = self.eval_domain.evaluate(&evals, r);
+
+            claimed_evaluation = reduced_evaluation;
+            evaluation_point.push(r);
+        }
+
+        if let Some(openings_claim) = openings_claim {
+            if openings_claim.evaluation_point != evaluation_point {
+                return Err(Error::WrongOpeningPoint);
+            }
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:112]: Is this check necessary? The prover can just put anything in `openings_claim.evaluation_point`, unrelated to the openings. So IIUC, we could just remove the field `OpeningClaims::evaluation_point`.

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            let query = self.final_query_builder.build_query(&openings_claim, &evaluation_point);
+            if self.composition_poly.evaluate(&query) != claimed_evaluation {
+                Err(Error::FinalEvaluationCheckFailed)
+            } else {
+                Ok(openings_claim)
+            }
+        } else {
+            Err(Error::NoOpeningsProvided)
+        }
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:121]: If it's always an error not to provide the openings, we should make the type of `Proof::openings` to be just `FinalOpeningClaim<E>` (i.e. *not* an `Option`)

**Review by plafer — COMMENTED** [review]
Looks good :) The documentation is very well done!

Not a full review yet

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs (hunk 2 above)
Comment (plafer) [reply processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:121]: I just noticed the comment on `Proof`:

```rs
/// Openings is an [Option] as there are situations where we would like to run the sum-check
/// protocol for a certain number of rounds that is less than the number of variables of the
/// multi-linears. This is the case for example when we have a merged polynomial.
```

What is meant by a "merged polynomial"?

File: processor/Cargo.toml [hunk 3]
```diff
@@ -23,9 +23,10 @@ internals = ["miden-air/internals"]
 std = ["vm-core/std", "winter-prover/std"]
 
 [dependencies]
+thiserror = { version = "1.0", default-features = false }
```
[the remark this answers is not in the document]
Comment (bitwalker) [reply processor/Cargo.toml:26]: Once #1277 is merged, we will have a version of `thiserror` (and `miette`) that work in no-std environments, using forks that I'm maintaining. Once `error_in_core` is stabilized, those forks can go away and we can use the mainline crates, but at least we'll have proper error infrastructure to lean on in the meantime

File: processor/src/trace/virtual_bus/sum_check/domain.rs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    pub fn evaluate(&self, evaluations: &[E], r: E) -> E {
+        evaluate_barycentric(&self.interpolation_points, evaluations, r, &self.barycentric_weights)
+    }
+}
+
+/// Computes the barycentric weights for a set of interpolation points.
+pub fn barycentric_weights<E: FieldElement>(points: &[E]) -> Vec<E> {
+    let n = points.len();
+    let tmp = (0..n)
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/domain.rs:37]: nit: would rename to `weight_denominators` or something of the like

File: processor/src/trace/virtual_bus/sum_check/domain.rs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        .map(|i| (0..n).filter(|&j| j != i).fold(E::ONE, |acc, j| acc * (points[i] - points[j])))
+        .collect::<Vec<_>>();
+    batch_inversion(&tmp)
+}
+
+/// Computes the value of the polynomial with minimal degree interpolating the set `{(x_i, y_i)}`
+/// at the point `r` given pre-computed barycentric weights.
+pub fn evaluate_barycentric<E: FieldElement>(
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/domain.rs:45]: I would either make `barycentric_weights()` and `evaluate_barycentric()` private, or put them in some `barycentric` utils module - this module should be only for sum-check domain.

For now though, I'd probably just make them private, as using `EvaluationDomain` will always be cleaner than using these 2 explicitly. And I'd also just inline `evaluate_barycentric()`

File: processor/src/trace/virtual_bus/sum_check/domain.rs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    x_i: &[E],
+    y_i: &[E],
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/domain.rs:47]: nit: I would rename to `xs` and `ys`, since `x_i` usually stands for "the ith `x`", and it sort of clashes with the iteration variable `i` in `x_i[i]` (and similarly for `y_i`)

File: processor/src/trace/virtual_bus/sum_check/domain.rs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        if x_i == r {
+            return y_i;
+        }
+    }
+
+    let l_x: E = x_i.iter().fold(E::ONE, |acc, &x_i| acc * (r - x_i));
+
+    let sum = (0..x_i.len()).fold(E::ZERO, |acc, i| {
+        let w_i = barycentric_weights[i];
+        acc + (w_i / (r - x_i[i]) * y_i[i])
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/domain.rs:61]:
```suggestion
        let term = (w_i / (r - x_i[i])) * y_i[i];
        acc + term
```

nit: I would add the extra parentheses so as to not rely on precedence rules which are easy to forget/get wrong. But this does become a bit paren-heavy, so I split it over 2 lines (feel free to re-arrange)

Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/domain.rs:61]: Is it worth using `batch_inversion()` here on the term `(r - x_i[i])` so as to avoid dividing `n` times?

File: processor/src/trace/virtual_bus/sum_check/mod.rs [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+        // Verifier
+        let plain_query_builder = PlainQueryBuilder;
+        let verifier = SumCheckVerifier::new(virtual_poly, plain_query_builder);
+        let mut coin = RpoRandomCoin::new(Word::default());
+        let result = verifier.verify(claim, proof, &mut coin);
+
+        assert!(result.is_err())
+    }
+
+    struct PlainQueryBuilder;
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/mod.rs:208]:
```suggestion
    struct ProjectionQueryBuilder;
```

Since this is the case where `f_i` returns the `i`th input

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+/// `v_i = f_i(r_0, ... ,r_{\nu - 1})`, this is the exact content of [FinalOpeningClaim], which can
+/// be either answered by a direct query to the oracles (i.e., in the complied protocol this would
+/// be answered with an opening proof against the commitment) or through further interaction (as
+/// in the case of the GKR protocol).
+///
+/// The purpose of [FinalQueryBuilder] is to abstract the logic for evaluating multi-linears which
+/// the Verifier can do by herself. For example, this is the case for periodic columns where given
+/// `(r_0, ... ,r_{\nu - 1})` the Verifier can evaluate `f_i(r_0, ... ,r_{\nu - 1})` un-assisted.
+///
+/// In the case where there are no `f_i(r_0, ... ,r_{\nu - 1})` which can be computed by the
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:142]: This is a bit misleading - we always have `f_i`s. The case being described here is the case where `f_i` returns the `i`th input. Saying that there are no `f_i` contradicts the initial problem statement, and is a bit confusing. I would just leave this out

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+        } else {
+            Err(Error::NoOpeningsProvided)
+        }
+    }
+}
+
+/// Contains the logic for building the final query made to the virtual polynomial.
+///
+/// During the last step of the sum-check protocol, the Verifier must evaluate the virtual
+/// polynomial at a random point `(r_0, ... ,r_{\nu - 1})`. To do this, the Verifier asks the Prover
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:128]: The term "virtual polynomial" doesn't appear in the `SumCheckProver` docstring that we use as central point of documentation of the algorithm. Could we clearly document what is meant by "virtual polynomial" there?

File: processor/src/trace/virtual_bus/sum_check/prover/mod.rs [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+///        (1 - X_i) . f_i(0, x_{i + 1}, \cdots, x_{\nu - 1}) + X_i . f_i(1, x_{i + 1}, \cdots, x_{\nu - 1})
+///
+/// Note that we omitted writing the folding randomness for readability.
+/// Since the evaluation domain is {0, 1, ... , d_max}, we can compute the evaluations based on
+/// the previous one using only additions. This is the purpose of `deltas`, to hold the increments
+/// added to each multi-linear to compute the evaluation at the next point, and `evals_x` to hold
+/// the current evaluation at `x` in {2, ... , d_max}.
+fn sumcheck_round<E: FieldElement>(
+    polynomial: &dyn CompositionPolynomial<E>,
+    mls: &mut [MultiLinear<E>],
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/prover/mod.rs:244]:
```suggestion
    mls: &[MultiLinear<E>],
```

File: processor/src/trace/virtual_bus/sum_check/prover/mod.rs [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+        let mut current_round_claim = RoundClaim {
+            eval_point: vec![],
+            claim,
+        };
+
+        // run the first round of the protocol
+        let mut round_proof = sumcheck_round(&self.composition_poly, mls);
+        round_proofs.push(round_proof);
+        // reseed with the s_0 polynomial
+        coin.reseed(H::hash_elements(&round_proofs[0].poly_evals));
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/prover/mod.rs:156]: nit: I would first reseed the coin using `&round_proof.poly_evals`, and then push `round_proof` in `round_proofs`

File: processor/src/trace/virtual_bus/sum_check/prover/mod.rs [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - No multi-linears were provided.
+    /// - Number of rounds is zero or is greater than the number of variables of the multilinears.
+    /// - The provided multi-linears have different arities.
+    pub fn prove(
+        &self,
+        claim: E,
+        mls: &mut [MultiLinear<E>],
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/prover/mod.rs:113]: We have a name for `g()`: the "composition polynomial". Could we also have a name for the `f_i`s, that's more descriptive than "multilinears"?

File: processor/src/trace/virtual_bus/sum_check/prover/mod.rs [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+            // run the i-th round of the protocol using the folded multi-linears for the new reduced
+            // claim. This basically computes the s_i polynomial.
+            round_proof = sumcheck_round(&self.composition_poly, mls);
+            round_proofs.push(round_proof);
+
+            // update the claim
+            current_round_claim = new_round_claim;
+
+            // reseed with the s_i polynomial
+            coin.reseed(H::hash_elements(&round_proofs[i].poly_evals));
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/prover/mod.rs:182]: Similarly, I would reseed the coin on `round_proof`

File: processor/src/trace/virtual_bus/multilinear/mod.rs [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    /// (1 - r_0) * f(0, y_1, ..., y_{ν - 1}) + r_0 * f(1, y_1, ..., y_{ν - 1}) and assigns
+    /// the resulting multi-linear, defined over a domain of half the size, to `self`.
+    pub fn bind_assign(&mut self, round_challenge: E) {
+        let mut result = vec![E::ZERO; 1 << (self.num_variables() - 1)];
+        for (i, res) in result.iter_mut().enumerate() {
+            *res = self.evaluations[i << 1]
+                + round_challenge * (self.evaluations[(i << 1) + 1] - self.evaluations[i << 1]);
+        }
+        *self =
+            Self::from_values(&result).expect("should not fail given that it is a multi-linear");
```
Comment (plafer) [inline processor/src/trace/virtual_bus/multilinear/mod.rs:93]: `bind_assign(&mut self)` to me suggests that this is done in-place, and that there is a performance benefit over using `bind()`. Instead, this is a new function for `*self = self.bind(...)`. I would remove this function and use `bind()` everywhere instead

File: processor/src/trace/virtual_bus/sum_check/prover/mod.rs [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+            // compute the new reduced round claim
+            let new_round_claim = reduce_claim(
+                &self.eval_domain,
+                &round_proofs[i - 1],
+                current_round_claim,
+                round_challenge,
+            );
+
+            // fold each multi-linear using the round challenge
+            mls.iter_mut().for_each(|ml| ml.bind_assign(round_challenge));
```
Comment (plafer) [inline processor/src/trace/virtual_bus/sum_check/prover/mod.rs:171]: Question: are there benefits to binding on each iteration, as opposed to keeping the initial multilinears as-is, and prepending the previous rounds challenges `[r0, ..., rm]` to the input? I'm not sure of the relative costs of binding vs running `sumcheck_round` on more variables.

File: processor/src/trace/virtual_bus/multilinear/mod.rs [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+fn inner_product<E: FieldElement>(evaluations: &[E], tensored_query: &[E]) -> E {
+    assert_eq!(evaluations.len(), tensored_query.len());
+    evaluations
+        .iter()
+        .zip(tensored_query.iter())
+        .fold(E::ZERO, |acc, (x_i, y_i)| acc + *x_i * *y_i)
+}
+
+/// Computes the evaluations of the Lagrange basis polynomials over the interpolating
+/// set {0 , 1}^ν at (r_0, ..., r_{ν - 1}) i.e., the Lagrange kernel at (r_0, ..., r_{ν - 1}).
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/multilinear/mod.rs:135]: I think this is used only in few places, so it shouldn't be that taxing in my opinion

File: processor/src/trace/virtual_bus/multilinear/mod.rs [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+pub fn tensorize<E: FieldElement>(query: &[E]) -> Vec<E> {
+    let nu = query.len();
+    let n = 1 << nu;
+
+    let mut evals: Vec<E> = vec![E::ONE; n];
+    let mut size = 1;
+    for query in query.iter() {
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/multilinear/mod.rs:142]: Changed to `r_i`  in the body and left `query` as it is more descriptive

File: processor/src/trace/virtual_bus/multilinear/mod.rs [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+/// The evaluations of a mult-linear polynomial over the boolean hyper-cube {0 , 1}^ν.
+#[derive(Clone, Debug)]
+pub struct MultiLinear<E: FieldElement> {
+    num_variables: usize,
+    evaluations: Vec<E>,
+}
+
+impl<E: FieldElement> MultiLinear<E> {
+    /// Constructs a [MultiLinear] from an array of values in the field.
+    pub fn new(values: Vec<E>) -> Result<Self, Error> {
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/multilinear/mod.rs:24]: Agreed and changed

File: processor/src/trace/virtual_bus/multilinear/mod.rs [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+            return Err(Error::EvaluationsNotPowerOfTwo);
+        }
+        Ok(Self {
+            num_variables: log2(values.len()) as usize,
+            evaluations: values,
+        })
+    }
+
+    /// Constructs a [MultiLinear] from a slice of values in the field.
+    pub fn from_values(values: &[E]) -> Result<Self, Error> {
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/multilinear/mod.rs:35]: Done

File: processor/src/trace/virtual_bus/multilinear/mod.rs [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+
+mod error;
+use self::error::Error;
+
+// MULTI-LINEAR POLYNOMIAL
+// ================================================================================================
+
+/// The evaluations of a mult-linear polynomial over the boolean hyper-cube {0 , 1}^ν.
+#[derive(Clone, Debug)]
+pub struct MultiLinear<E: FieldElement> {
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/multilinear/mod.rs:17]: Changed

File: processor/src/trace/virtual_bus/multilinear/mod.rs [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+    assert_eq!(evaluations.len(), tensored_query.len());
+    evaluations
+        .iter()
+        .zip(tensored_query.iter())
+        .fold(E::ZERO, |acc, (x_i, y_i)| acc + *x_i * *y_i)
+}
+
+/// Computes the evaluations of the Lagrange basis polynomials over the interpolating
+/// set {0 , 1}^ν at (r_0, ..., r_{ν - 1}) i.e., the Lagrange kernel at (r_0, ..., r_{ν - 1}).
+pub fn tensorize<E: FieldElement>(query: &[E]) -> Vec<E> {
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/multilinear/mod.rs:136]: Will do that in the next PR. 
Will leave this conversation here so that I remember

File: processor/src/trace/virtual_bus/sum_check/prover/mod.rs [hunk 23]
```diff
@@ -0,0 +1,289 @@
...
+use vm_core::{FieldElement, StarkField};
+use winter_prover::crypto::{ElementHasher, RandomCoin};
+
+mod error;
+
+/// A struct that contains relevant information for the execution of the multivariate sum-check
+/// protocol prover.
+/// The sum-check protocol is an interactive protocol (IP) for proving the following relation:
+///
+/// v = \sum_{(x_0,\cdots, x_{\nu - 1}) \in \{0 , 1\}^{2^{\nu}}}
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/prover/mod.rs:16]: Good catch

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs (hunk 1 above)
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:112]: You are right but there are two reasons for having it as part of the claim:

1. In the case where we have many different runs of different sum-checks, we would have different opening points against some of the same oracles. At this point, one can usually do a batch opening argument in order to reduce the cost of the opening proof. This is probably not immediately relevant for us but could be in the future.
2. The more relevant reason for us is so that the STARK prover would have all the data necessary in order to build the auxiliary trace helper columns in order to do the (multi-linear) opening proof.

To summarize, we could, and probably, should remove the evaluation points (in order to reduce the size of the GKR proofs), but having it this way is convenient at the moment and potentially necessary in the future.

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs (hunk 2 above)
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:121]: > If it's always an error not to provide the openings, we should make the type of `Proof::openings` to be just `FinalOpeningClaim<E>` (i.e. _not_ an `Option`)

I have changed this in the subsequent PR as this was a poor design choice

Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:121]: > I just noticed the comment on `Proof`:
> 
> ```rust
> /// Openings is an [Option] as there are situations where we would like to run the sum-check
> /// protocol for a certain number of rounds that is less than the number of variables of the
> /// multi-linears. This is the case for example when we have a merged polynomial.
> ```
> 
> What is meant by a "merged polynomial"?

Fixed now

File: processor/src/trace/virtual_bus/multilinear/mod.rs [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+// COMPOSITION POLYNOMIAL
+// ================================================================================================
+
+/// A multi-variate polynomial for composing individual multi-linear polynomials.
+pub trait CompositionPolynomial<E: FieldElement>: Sync + Send {
+    /// The number of variables when interpreted as a multi-variate polynomial.
+    fn num_variables(&self) -> usize;
+
+    /// Maximum degree in all variables.
+    fn max_degree(&self) -> usize;
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/multilinear/mod.rs:114]: Changed now

File: processor/src/trace/virtual_bus/sum_check/domain.rs (hunk 4 above)
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/domain.rs:37]: changed

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+        claim: E,
+        proof: Proof<E>,
+        coin: &mut C,
+    ) -> Result<FinalOpeningClaim<E>, Error> {
+        let Proof {
+            openings: openings_claim,
+            round_proofs,
+        } = proof;
+
+        let mut claimed_evaluation = claim;
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:95]: Changed

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs [hunk 26]
```diff
@@ -0,0 +1,152 @@
...
+/// be answered with an opening proof against the commitment) or through further interaction (as
+/// in the case of the GKR protocol).
+///
+/// The purpose of [FinalQueryBuilder] is to abstract the logic for evaluating multi-linears which
+/// the Verifier can do by herself. For example, this is the case for periodic columns where given
+/// `(r_0, ... ,r_{\nu - 1})` the Verifier can evaluate `f_i(r_0, ... ,r_{\nu - 1})` un-assisted.
+///
+/// In the case where there are no `f_i(r_0, ... ,r_{\nu - 1})` which can be computed by the
+/// Verifier alone, the output of [Self::build_query] will be just the provided openings.
+pub trait FinalQueryBuilder {
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:144]: Changed

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+    type Field: FieldElement;
+
+    fn build_query(
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:147]: Changed

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs [hunk 28]
```diff
@@ -0,0 +1,152 @@
...
+/// 1. Multi-round interaction where it sends challenges and receives polynomials. For each
+/// polynomial received it uses the sent randomness to reduce the current claim to a new one.
+///
+/// 2. A final round where the Verifier queries the multi-linear oracles it received at the outset
+/// of the protocol (i.e., commitments) for their evaluations the random point
+/// `(r_0, ... , r_{\nu - 1})` where $\nu$ is the number of rounds of the sum-check protocol and
+/// `r_i` is the randomness sent by the Verifier at each round.
+pub struct SumCheckVerifier<B, E, P, C, H, V>
+where
+    B: StarkField,
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:24]: Changed

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs (hunk 9 above)
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:142]: Removed

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs (hunk 10 above)
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/verifier/mod.rs:128]: Updated

File: processor/src/trace/virtual_bus/sum_check/prover/mod.rs [hunk 29]
```diff
@@ -0,0 +1,289 @@
...
+/// 2. The Prover has each `f_i` in its evaluation form over the hyper-cube \{0 , 1\}^{2^{\nu}}.
+///
+/// 3. An optimization is for the Prover to not send `s_i(0)` as it can be recoverd from the current
+/// reduced claim s_{i - 1}(r_{i - 1}) using the relation s_{i}(0) = s_{i}(1) - s_{i - 1}(r_{i - 1}).
+/// This also means that the Verifier can skip point 4.b.
+pub struct SumCheckProver<B, E, P, C, H>
+where
+    B: StarkField,
+    E: FieldElement<BaseField = B>,
+    C: RandomCoin<Hasher = H, BaseField = B>,
```
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/prover/mod.rs:71]: But then would have the same issue with `H`?

File: processor/src/trace/virtual_bus/sum_check/prover/mod.rs (hunk 12 above)
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/prover/mod.rs:156]: Makes sense

File: processor/src/trace/virtual_bus/sum_check/prover/mod.rs (hunk 13 above)
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/prover/mod.rs:113]: We could use: multi-linear extension polynomials but I am not sure this is more descriptive

File: processor/src/trace/virtual_bus/multilinear/mod.rs (hunk 15 above)
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/multilinear/mod.rs:93]: I would like it to be in place but not sure about the most straightforward way to do this
I removed and renamed as suggested

File: processor/src/trace/virtual_bus/sum_check/prover/mod.rs (hunk 16 above)
Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/prover/mod.rs:171]: I am not sure about the meaning of prepending the previous round challenges

File: processor/Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Al-Kindi-0) [reply processor/Cargo.toml:26]: > Once #1277 is merged, we will have a version of `thiserror` (and `miette`) that work in no-std environments, using forks that I'm maintaining. Once `error_in_core` is stabilized, those forks can go away and we can use the mainline crates, but at least we'll have proper error infrastructure to lean on in the meantime

Thank you Paul

**Comment by Al-Kindi-0** [comment]
I have addressed most of the feedback but for the ones related to the evaluation domain. The reason for this is that we will be moving to the coefficient in a subsequent PR and we can implement most of the feedback as part of that PR.

File: processor/src/trace/virtual_bus/sum_check/domain.rs (hunk 4 above)
Comment (plafer) [reply processor/src/trace/virtual_bus/sum_check/domain.rs:37]: Actually I was referring to the `tmp` variable to be renamed to `weight_denominators`, since it contains the denominators before they get inverted. I think `barycentric_weights()` is a good name for the function, since it agrees with the definition of barycentric weights [in the literature](https://people.maths.ox.ac.uk/trefethen/barycentric.pdf).

File: processor/src/trace/virtual_bus/sum_check/verifier/mod.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (plafer) [reply processor/src/trace/virtual_bus/sum_check/verifier/mod.rs]: The current change made `CompositionPolyQueryBuilder` generic over `E`. I think it would be better if `build_query()` was generic over `E` instead.

This is because `impl<E> CompositionPolyQueryBuilder<E> for X` results in multiple trait implementations (one per `E`). Hence you can no longer do `X.build_query()`, because the compiler no longer knows which `build_query()` you're referring to (except in [cases](https://github.com/0xPolygonMiden/miden-vm/blob/9764c43712fcb6f5b7a25d973c321116c992ea53/processor/src/trace/virtual_bus/sum_check/verifier/mod.rs#L101) where the implementation is also generic over that implementation). It would also remove the need for [this `PhantomData`](https://github.com/0xPolygonMiden/miden-vm/blob/9764c43712fcb6f5b7a25d973c321116c992ea53/processor/src/trace/virtual_bus/sum_check/mod.rs#L226).

File: processor/src/trace/virtual_bus/sum_check/prover/mod.rs (hunk 29 above)
[the remark this answers is not in the document]
Comment (plafer) [reply processor/src/trace/virtual_bus/sum_check/prover/mod.rs]: I'm no longer sure why I suggested this, looks fine to me as is.

File: processor/src/trace/virtual_bus/sum_check/prover/mod.rs (hunk 16 above)
Comment (plafer) [reply processor/src/trace/virtual_bus/sum_check/prover/mod.rs]: Basically, we'd leave the multilinears untouched, and call them e.g. as `ml.evaluate([r0, ..., rm, x])`, where `r0, ... rm` are the previous rounds' claims.

Basically, run the algorithm as described in the literature - keeping the `r_i`s around, and no binding.

Comment (Al-Kindi-0) [reply processor/src/trace/virtual_bus/sum_check/prover/mod.rs]: Ok, and what are we comparing here exactly?

> I'm not sure of the relative costs of binding vs running sumcheck_round on more variables.

**Comment by plafer** [comment]
Should we just close in PR in favor of #1307?

**Comment by Al-Kindi-0** [comment]
The implementation has now moved to using the coefficient form for the round polynomial instead of the evaluation one. With this change, we avoid having to deal with evaluation domains.
Also, regarding `CompositionPolyQueryBuilder`, I had to revert back to the previous implementation which had the trait generic over the field.  The reason been that for the following typical use case, the previous approach didn't work as (maybe there is a workaround I am not aware of)
```rust
/// A [`FinalQueryBuilder`] for the sum-check verifier used for the final sum-check.
#[derive(Default)]
struct GkrMergeQueryBuilder<E> {
    gkr_eval_point: Vec<E>,
    merge_rand: Vec<E>,
}

impl<E: FieldElement> CompositionPolyQueryBuilder<E> for GkrMergeQueryBuilder<E> {
    fn build_query(&self, openings_claim: &FinalOpeningClaim<E>, evaluation_point: &[E]) -> Vec<E> {
        let eq_at_gkr_eval_point = EqFunction::new(self.gkr_eval_point.clone());
        let mut rand_sumcheck = self.merge_rand.clone();
        rand_sumcheck.extend_from_slice(evaluation_point);
        let eq = eq_at_gkr_eval_point.evaluate(&rand_sumcheck);
        let mut query = openings_claim.openings.clone();
        query.push(eq);
        query
    }
}
```

As for closing this PR, I agree. I think that there no outstanding comments left here. The review for the change to coefficient form could be done in the other PR.

**Comment by plafer** [comment]
Closing; let's move all discussion to #1307

**Comment by plafer** [comment]
Reopening - @Al-Kindi-0 could you merge your recent commits into #1307, and then close this one?

**Comment by Al-Kindi-0** [comment]
> Reopening - @Al-Kindi-0 could you merge your recent commits into #1307, and then close this one?

Done! I will now close this.
